x-pack/plugins/security_solution/common/api/timeline/model/components.schema.yaml

```diff
@@ -301,15 +145,62 @@ components:
         - system_socket
         - threat_match
         - zeek
+    QueryMatchResult:
+      type: object
+      properties:
+        field:
+          type: string
+        displayField:
+          type: string
+        value:
+          type: string
+        displayValue:
+          type: string
+        operator:
+          type: string
+    BaseDataProviderResult:
+      type: object
+      properties:
+        id:
+          type: string
+        name:
+          type: string
+        enabled:
+          type: boolean
+        excluded:
+          type: boolean
+        kqlQuery:
+          type: string
+        queryMatch:
+          $ref: '#/components/schemas/QueryMatchResult'
+        type:
+          $ref: '#/components/schemas/DataProviderType'
+    DataProviderResult:
+      x-recursive: true
+      x-recursive-prop: and
+      x-recursive-value: DataProviderResult[]
+      x-base-type: BaseDataProviderResult
```

Comment:
This is a fair amount of different properties that must be specified correctly to make the TS type for the recursive schema, and it's pretty specific to this particular case. E.g. what if the recursive prop was 2 levels deep, or we wanted to combine multiple base types. It's also a bit worrisome to me that `x-recursive-value` stores an actual TS type to be injected into the generated code, as opposed to being a more abstract control value that affects the code generation. Going in that direction, mixing TS and OpenAPI specs, could make it harder to maintain in the future.

Have you seen @xcrzx [PoC for generating TS types](https://github.com/elastic/kibana/pull/168813/files#diff-f719628b17adf1fd07e2f660a769bf6d9650b419fee0c636ad9041d86a7ecf1c) directly from the OpenAPI schema? I wonder if recursive schemas would benefit from polishing up the PoC and making it available for use. A flag like `x-typegen-enabled` could trigger the codegen to replace `z.infer<...>` with a TS type generated directly from the OpenAPI schema. Generating the type directly would hopefully eliminate the need to add the additional `x-` configuration hints to create a TS type.

Reply:
I didn't see it, but will take a look! Yea, this solution is not ideal by any means, just a first pass...was hoping to get feedback on it.